### PR TITLE
CDAP-8014 Cache RouteDestination as an optimization

### PR DESCRIPTION
**Pull request by Denton-L**
JIRA: https://issues.cask.co/browse/CDAP-8014
Bamboo: http://builds.cask.co/browse/CDAP-DUT5249

### CODE REVIEW

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 1]
```diff
@@ -41,6 +46,9 @@
   public static final RouteDestination STREAMS_SERVICE = new RouteDestination(Constants.Service.STREAMS);
   public static final RouteDestination PREVIEW_HTTP = new RouteDestination(Constants.Service.PREVIEW_HTTP);
 
+  private Map<MultiKey, RouteDestination> _routeDestinationCache = Collections.synchronizedMap(
```
Comment (sreevatsanraman) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: remove the underscore prefix in the variable name.

Comment (Denton-L) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: Should we remove item 15 from the coding standards page on Confluence?

**Comment by Denton-L** [comment]
Should we remove item 15 from the coding standards page on Confluence?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 2]
```diff
@@ -41,6 +46,9 @@
   public static final RouteDestination STREAMS_SERVICE = new RouteDestination(Constants.Service.STREAMS);
   public static final RouteDestination PREVIEW_HTTP = new RouteDestination(Constants.Service.PREVIEW_HTTP);
 
+  private Map<MultiKey, RouteDestination> routeDestinationCache = Collections.synchronizedMap(
```
Comment (bdmogal) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: final

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 3]
```diff
@@ -83,8 +91,12 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
       //       <user-defined-method-path>"
       String serviceName = ServiceDiscoverable.getName(uriParts[2], uriParts[4], uriParts[8]);
       String version = uriParts[6];
-      // TODO: Optimize this instead of creating new object every time
-      return new RouteDestination(serviceName, version);
+      MultiKey cacheKey = new MultiKey(serviceName, version);
+
+      if (!routeDestinationCache.containsKey(cacheKey)) {
```
Comment (bdmogal) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:94]: hmmm, why do you need a synchronized map? Also, if you do a ``containsKey`` + ``put``, its not really synchronized. Use ``putIfAbsent``.

**Comment by Denton-L** [comment]
@bdmogal I switched to ConcurrentMap but, unfortunately, I cannot use `putIfAbsent` because this would cause an object to be created anyway which is exactly what we are trying to prevent. 

EDIT: Realistically, I don't think that there would be any synchronization problems. Even if the "wrong" object were placed in the map, it would still work as expected, I believe.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 4]
```diff
@@ -83,8 +89,12 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
       //       <user-defined-method-path>"
       String serviceName = ServiceDiscoverable.getName(uriParts[2], uriParts[4], uriParts[8]);
       String version = uriParts[6];
-      // TODO: Optimize this instead of creating new object every time
-      return new RouteDestination(serviceName, version);
+      MultiKey cacheKey = new MultiKey(serviceName, version);
```
Comment (chtyim) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:92]: This is not an optimization at all. It used to creating one `RouteDestination` object. Now it is creation of one `MultiKey`, plus two lookup in a concurrent map (`containsKey` and `get`). I don't think we need to change anything here. Simply remove the TODO and mark the CDAP-8014 as won't fix.

Shouldn't do any optimization without actual perf test backing it.

**Pull request closed without merging**